File: python-sdk/src/mlxmesh/pow.py
"""Proof-of-work nonce mining for POST /users/{id}/startup-grant.

Matches internal/settlement/pow.go's VerifyProofOfWork exactly: the
coordinator requires a uint64 nonce such that sha256(user_id_bytes ++
big_endian_uint64(nonce)) has >= difficulty_bits leading zero BITS (not
bytes). DefaultGrantPoWBits = 18 averages ~262k hashes — sub-second on real
hardware — cheap enough not to annoy a real claim, expensive enough to make
Sybil-farming disposable user_ids cost real wall-clock time.
"""

from __future__ import annotations

import hashlib

DEFAULT_GRANT_POW_BITS = 18


def _leading_zero_bits(digest: bytes) -> int:
    count = 0
    for byte in digest:
        if byte == 0:
            count += 8
            continue
        mask = 0x80
        while mask:
            if byte & mask:
                return count
            count += 1
            mask >>= 1
    return count


def verify_proof_of_work(user_id: str, nonce: int, difficulty_bits: int) -> bool:
    """Pure check, mirroring the Go side — used by mine() and by tests to
    confirm interop without needing a live coordinator."""
    if difficulty_bits <= 0:
        return True
    digest = hashlib.sha256(user_id.encode("utf-8") + nonce.to_bytes(8, "big")).digest()
    return _leading_zero_bits(digest) >= difficulty_bits


def mine(user_id: str, difficulty_bits: int = DEFAULT_GRANT_POW_BITS) -> int:
    """Brute-forces the smallest nonce satisfying verify_proof_of_work.
    Deterministic starting point (0) — anyone re-mining the same user_id at
    the same difficulty gets the same nonce, which is fine: the coordinator
    doesn't care which valid nonce you submit, only that one is valid, and a
    claim is idempotent (a second claim is a no-op, not an error)."""
    nonce = 0
    while not verify_proof_of_work(user_id, nonce, difficulty_bits):
        nonce += 1
    return nonce

File: swift-sdk/Sources/MeshKit/PoW.swift
import CryptoKit
import Foundation

/// Proof-of-work nonce mining for POST /users/{id}/startup-grant. Matches
/// internal/settlement/pow.go's VerifyProofOfWork exactly: the coordinator
/// requires a UInt64 nonce such that sha256(user_id_bytes ++
/// big_endian_uint64(nonce)) has >= difficultyBits leading zero BITS (not
/// bytes). The default of 18 bits averages ~262k hashes — sub-second on real
/// hardware.
public enum ProofOfWork {
    public static let defaultGrantPoWBits = 18

    static func leadingZeroBits(_ digest: SHA256.Digest) -> Int {
        var count = 0
        for byte in digest {
            if byte == 0 {
                count += 8
                continue
            }
            var mask: UInt8 = 0x80
            while mask != 0 {
                if byte & mask != 0 { return count }
                count += 1
                mask >>= 1
            }
        }
        return count
    }

    /// Pure check, mirroring the Go side — used by mine(_:difficultyBits:)
    /// and exposed for interop testing.
    public static func verify(userID: String, nonce: UInt64, difficultyBits: Int) -> Bool {
        if difficultyBits <= 0 { return true }
        var data = Data(userID.utf8)
        withUnsafeBytes(of: nonce.bigEndian) { data.append(contentsOf: $0) }
        let digest = SHA256.hash(data: data)
        return leadingZeroBits(digest) >= difficultyBits
    }

    /// Brute-forces the smallest nonce satisfying `verify`. Deterministic
    /// starting point (0) — a claim is idempotent server-side, so it doesn't
    /// matter which valid nonce is submitted.
    public static func mine(userID: String, difficultyBits: Int = defaultGrantPoWBits) -> UInt64 {
        var nonce: UInt64 = 0
        while !verify(userID: userID, nonce: nonce, difficultyBits: difficultyBits) {
            nonce += 1
        }
        return nonce
    }
}
